### ISSUE TITLE
Messages from background threads are not logged to session log

### ISSUE THREAD

**Issue by bpowell65536** [getslash/slash#949]
When running the code given below with `slash run test.py`, the messages logged by the `BackgroundLogger` object are written only to the test log file during a test, not to the session log file. When the test session is interrupted with Ctrl+C, the messages logged by the `BackgroundLogger` object show up in the session log file during the cleanup phase.

With `slash.config.root.log.unified_session_log = True` in .slashrc, I would expect the `BackgroundLogger` messages to always show up in the session log, even when a test is in progress. Have I misunderstood how this feature works, or is this a bug?

test.py:
```python
import logbook
import slash
import threading
import time


class BackgroundLogger(object):
    def __init__(self):
        self._running = threading.Event()
        self._thread = threading.Thread(target=self._run)
        self._thread.start()
        self._log = logbook.Logger("BackgroundLogger")

    def _run(self):
        thread_local_log = logbook.Logger("BackgroundLogger_thread")
        self._running.set()
        while self._running.is_set():
            thread_local_log.trace("")
            thread_local_log.debug("")
            thread_local_log.info("")
            thread_local_log.notice("")
            self._log.trace("")
            self._log.debug("")
            self._log.info("")
            self._log.notice("")
            time.sleep(1)

    def stop(self):
        # Wait for the thread to log some messages during cleanup
        time.sleep(10)
        self._running.clear()
        self._thread.join()


@slash.fixture(scope="session")
def background_logger():
    logger = BackgroundLogger()
    slash.add_critical_cleanup(logger.stop)


@slash.parametrize("dummy", range(10))
def test_dummy(background_logger, dummy):
    log = logbook.Logger("ForegroundLogger")
    for _ in range(5):
        log.trace("")
        log.debug("")
        log.info("")
        log.notice("")
        time.sleep(1)
```

.slashrc:
```python
import logbook
import slash


slash.config.root.log.root = "logs"
slash.config.root.log.last_session_symlink = "../last_session"
slash.config.root.log.last_test_symlink = "../last_test"
slash.config.root.log.truncate_console_lines = False
slash.config.root.log.localtime = True
slash.config.root.log.core_log_level = logbook.TRACE
slash.config.root.log.console_level = logbook.NOTICE
slash.config.root.log.unified_session_log = True
slash.config.root.log.colorize = True
```

**Comment by vmalloc** [comment]
Hi! Thanks for reaching out!

Logging in Slash is application bound by default, and once a test starts all logging goes to the test log. This is because logically, everything that happens inside the test's duration is logically a part of it by default.

If you need your thread to log to the session log (or even better yet, a separate log file dedicated to its records) I suggest setting up a thread-bound handler upon the thread's start that emits to your desired location.

**Comment by bpowell65536** [comment]
Thanks for the quick response! I tried adding a thread-bound handler logging to the session log file by adding

```python
handler = slash.log.ColorizedFileHandler(
    filename=slash.session.logging.session_log_path,
    format_string=slash.config.root.log.format
)
handler.push_thread()
```

to the start of the `BackgroundLogger._run()` method, and now it logs messages from the thread to the session log, but only for the first test...

Is the thread/application binding specific to the file loggers here? The console logger seems to log messages from both the main thread and the background thread, which is what confused me for a start (sorry, I probably should have mentioned that).

**Comment by vmalloc** [comment]
@bpowell65536 yeah - now that I think of it we need to tweak the behavior a bit. I think the session-wide logging context should be application-bound, but then subsequent tests should be thread-bound. This is a bit tricky - we need to test that threads spawned from within the tests log correctly though (needs to be added in the PR that fixes this).

@ayalash / @maormarcus do you object to such a change? I don't see much harm that can be done assuming the above invariant is kept...

**Comment by ayalash** [comment]
@vmalloc Just to make sure I understood correctly the expected behavior, in case we'll do this change: When running without unified-session log, will the log messages which are written by non-main threads be written only to the session log?

If so, it is problematic for threads which are used as keep-alive threads (like backslash-python) and that we need to see their messages in the right context (eg. couldn't reach server on time)

**Comment by vmalloc** [comment]
Hmmm... that's a valid point. I guess we need some sort of mechanism for opting out of this behavior for threads who don't want their logs following through to the current test constantly... any ideas?

**Comment by ayalash** [comment]
@vmalloc I think I'm missing something...
As I see it, when you are without unified-session, everything is working as expected
When you are running with unified-session, I think that the expectation should be that the log line will appear in both log files (session and test) and not only the session one.  Why do we want it to appear (in unified-session mode) only in the session log?

**Comment by vmalloc** [comment]
@ayalash the goal here is to be able to create a thread that pushes its own handler and keeps it throughout the session. This is regardless of the unified session log... Currently if I'm understanding this correctly, it doesn't work as the test handlers that keep getting pushed override the thread-specific handler...

**Comment by JavierCorado** [comment]
Following up this discussion with a similar topic.

Can slash logger (logbook) capture logs from a test case that uses external libraries that use the python standard logging module?

I can see the external library output in the console but not in the log.


Is this something that should be discussed in a separate issue?
